tty/uart16550/uarthw-tda4vm.c

```diff
@@ -0,0 +1,284 @@
+#define MAIN_UART8_BASE_ADDR ((void *)0x02880000)
+#define MAIN_UART9_BASE_ADDR ((void *)0x02890000)
+
+#ifndef MAIN_UART0_IRQ
+#define MAIN_UART0_IRQ -1
+#endif
+
+#ifndef MAIN_UART1_IRQ
+#define MAIN_UART1_IRQ -1
+#endif
+
+#ifndef MAIN_UART2_IRQ
+#define MAIN_UART2_IRQ -1
+#endif
+
+#ifndef MAIN_UART3_IRQ
+#define MAIN_UART3_IRQ -1
+#endif
+
+#ifndef MAIN_UART4_IRQ
+#define MAIN_UART4_IRQ -1
+#endif
+
+#ifndef MAIN_UART5_IRQ
+#define MAIN_UART5_IRQ -1
+#endif
+
+#ifndef MAIN_UART6_IRQ
+#define MAIN_UART6_IRQ -1
+#endif
+
+#ifndef MAIN_UART7_IRQ
+#define MAIN_UART7_IRQ -1
+#endif
+
+#ifndef MAIN_UART8_IRQ
+#define MAIN_UART8_IRQ -1
+#endif
+
+#ifndef MAIN_UART9_IRQ
+#define MAIN_UART9_IRQ -1
+#endif
+
+
+#define MAX_PINS_PER_UART 6
+typedef struct {
+	volatile uint32_t *base;
+	struct {
+		int16_t clksel; /* Value of < 0 means this UART has no CLKSEL */
+		uint8_t clksel_val;
+	};
+	int16_t clkdiv;
+	uint8_t divisor;
+	uint8_t pll;
+	uint8_t hsdiv;
+	int irq;
+	struct {
+		int16_t pin; /* Value of < 0 signals end of list */
+		uint8_t muxSetting;
+		uint8_t isTx;
+	} pins[MAX_PINS_PER_UART];
+	struct {
+		uint32_t tx;
+		uint32_t rx;
+	} pins_selected;
+} tda4vm_uart_info_t;
+
+
+static const tda4vm_uart_info_t uarthw_info[] = {
+	{
+		.base = MCU_UART0_BASE_ADDR,
+		.clksel = clksel_mcu_usart,
+		.clksel_val = 0, /* CLKSEL set to MCU_PLL1_HSDIV3_CLKOUT */
+		.clkdiv = -1,
+		.divisor = 1,
+		.pll = clk_mcu_per_pll1,
+		.hsdiv = 3,
+		.irq = MCU_UART0_IRQ,
+		.pins = {
+			{ pin_mcu_ospi1_d2, 4, 1 },
+			{ pin_wkup_gpio0_10, 2, 1 },
+			{ pin_wkup_gpio0_12, 0, 1 },
+			{ pin_mcu_ospi1_d1, 4, 0 },
+			{ pin_wkup_gpio0_11, 2, 0 },
+			{ pin_wkup_gpio0_13, 0, 0 },
+		},
+		.pins_selected = {
+			.tx = UART0_TX,
+			.rx = UART0_RX,
+		},
+	},
+};
+
+
+typedef struct {
+	volatile uint32_t *base;
+	unsigned int irq;
+} uarthw_ctx_t;
+
+
+uint8_t uarthw_read(void *hwctx, unsigned int reg)
+{
+	volatile uint32_t *p = (((uarthw_ctx_t *)hwctx)->base + reg);
+
+	return ((uint8_t)(*p));
+}
+
+
+void uarthw_write(void *hwctx, unsigned int reg, uint8_t val)
+{
+	volatile uint32_t *p = (((uarthw_ctx_t *)hwctx)->base + reg);
+
+	*p = val;
+}
+
+
+char *uarthw_dump(void *hwctx, char *s, size_t sz)
+{
+	snprintf(s, sz, "base=0x%p irq=%u", (volatile void *)(((uarthw_ctx_t *)hwctx)->base), ((uarthw_ctx_t *)hwctx)->irq);
+	return s;
+}
+
+
+unsigned int uarthw_irq(void *hwctx)
+{
+	return ((uarthw_ctx_t *)hwctx)->irq;
+}
+
+
+static int uarthw_setPin(const tda4vm_uart_info_t *info, uint32_t pin)
+{
+	int muxSetting, isTx;
+	platformctl_t pctl;
+	pctl.action = pctl_set;
+	pctl.type = pctl_pinconfig;
+
+	for (unsigned i = 0; i < MAX_PINS_PER_UART; i++) {
+		if (info->pins[i].pin < 0) {
+			return -EINVAL;
+		}
+
+		if (info->pins[i].pin == pin) {
+			muxSetting = info->pins[i].muxSetting;
+			isTx = info->pins[i].isTx;
+			pctl.pin_config.pin_num = pin;
+			pctl.pin_config.debounce_idx = 0;
+			pctl.pin_config.mux = muxSetting & 0xff;
```

Comment:
`info->pins[i].muxSetting` is defined as `uint8_t` - why store it in a temporary `int` variable if it can be assigned directly to `pctl.pin_config.mux` without a mask?